sound/soc/sof/debug.c

```diff
@@ -626,6 +627,124 @@ int snd_sof_debugfs_buf_item(struct snd_sof_dev *sdev,
 }
 EXPORT_SYMBOL_GPL(snd_sof_debugfs_buf_item);
 
+static int memory_info_update(struct snd_sof_dev *sdev, char **out)
+{
+	struct sof_ipc_cmd_hdr msg = {
+		.size = sizeof(struct sof_ipc_cmd_hdr),
+		.cmd = SOF_IPC_GLB_DEBUG | SOF_IPC_DEBUG_MEM_USAGE,
+	};
+	struct sof_ipc_dbg_mem_usage *reply;
+	int len = 0;
+	char *buf;
+	int ret;
+	int i;
+
+	reply = kmalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	if (!reply)
+		return -ENOMEM;
+
+	ret = pm_runtime_get_sync(sdev->dev);
+	if (ret < 0 && ret != -EACCES) {
+		pm_runtime_put_noidle(sdev->dev);
+		dev_err(sdev->dev, "error: enabling device failed: %d\n", ret);
+		ret = -EINVAL;
+		goto error;
+	}
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg.cmd, &msg, msg.size, reply, SOF_IPC_MSG_MAX_SIZE);
+	pm_runtime_mark_last_busy(sdev->dev);
+	pm_runtime_put_autosuspend(sdev->dev);
+	if (ret < 0 || reply->rhdr.error < 0) {
+		dev_err(sdev->dev, "error: reading memory info failed, %d\n",
+			min(ret, reply->rhdr.error));
+		ret = -EINVAL;
+		goto error;
+	}
+
+	if (struct_size(reply, elems, reply->num_elems) != reply->rhdr.hdr.size) {
+		dev_err(sdev->dev, "error: invalid memory info ipc struct size, %d\n",
+			reply->rhdr.hdr.size);
+		ret = -EINVAL;
+		goto error;
+	}
+
+	if (!buf) {
+		buf = devm_kmalloc(sdev->dev, PAGE_SIZE, GFP_KERNEL);
+		if (!buf) {
+			ret = -ENOMEM;
+			goto error;
+		}
+	}
+
+	for (i = 0; i < reply->num_elems; i++) {
+		ret = snprintf(buf + len, PAGE_SIZE - len, "zone %d.%d used %#8x free %#8x\n",
+			       reply->elems[i].zone, reply->elems[i].id,
+			       reply->elems[i].used, reply->elems[i].free);
+		if (ret < 0)
+			goto error;
+		len += ret;
+	}
+
+	*out = buf;
+	return len;
+error:
+	kfree(reply);
+	return ret;
+}
+
+static ssize_t memory_info_read(struct file *file, char __user *to, size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct snd_sof_dev *sdev = dfse->sdev;
+	static char *buf;
+	static int len;
+	int ret;
+
+	/* read memory info from FW only once for each file read */
+	if (!*ppos) {
+		len = memory_info_update(sdev, &buf);
+		if (len < 0)
+			return len;
+	}
+
+	ret = simple_read_from_buffer(to, count, ppos, buf, len);
+
+	return ret;
+}
+
+static const struct file_operations memory_info_fops = {
+	.open = simple_open,
+	.read = memory_info_read,
+	.llseek = default_llseek,
+};
+
+int snd_sof_dbg_memory_info_init(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_dfsentry *dfse;
+	size_t size = 512;
+
+	dfse = devm_kzalloc(sdev->dev, sizeof(*dfse), GFP_KERNEL);
+	if (!dfse)
+		goto error;
+
+	dfse->type = SOF_DFSENTRY_TYPE_BUF;
+	dfse->sdev = sdev;
+
+	dfse->size = size;
+	dfse->buf = devm_kmalloc(sdev->dev, size, GFP_KERNEL);
```

Comment:
didn't you want to use this buffer instead of the static one above?

Reply:
is it legal ?

Reply:
I don't see why it wouldn't be legal. `struct snd_sof_dfsentry` is SOF-specific, so, you can easily look through all uses of `.buf`- just search for `dfse->buf`. You don't call any of those functions for your debugfs file, so, `.buf` is all yours.

Reply:
Great, so I add also `buf_data_size` field to `struct snd_sof_dfsentry`, instead of keeping `memory_info_len` in `sdev`, because buffering should be available for each `dfse` read file